### PR TITLE
[Test][MOE] add MoeGroupedGemmNopadFwdOp test and bench and flip to implemented

### LINKED ISSUE #1474: [Test][MOE] add MoePermuteNopadFwdOp test and flip to implemented
### Symptom / Motivation

`MoePermuteNopadFwdOp` is `status: spec-only` despite having an Op class (`tileops/ops/moe/permute_nopad.py`), a kernel (`tileops/kernels/moe/permute_nopad.py`), and a dedicated bench (`benchmarks/ops/bench_moe_permute_nopad.py`). The blocking gap is **no test file**: the sibling `tests/ops/test_moe_permute.py` only exercises the padded variant.

### Root Cause Analysis

The nopad permute kernel landed without a parity test; the padded test file's coverage was not extended. `validate_manifest.py --strict --check-op MoePermuteNopadFwdOp` will reject the flip on test coverage (per workload conformance contract), and `source.kernel_map` is missing in the manifest entry.

### Related Files

- `tileops/manifest/moe.yaml` (entry `MoePermuteNopadFwdOp`)
- `tileops/ops/moe/permute_nopad.py`
- `tileops/kernels/moe/permute_nopad.py`
- `tests/ops/test_moe_permute.py` (target file to extend, or a sibling `test_moe_permute_nopad.py`)

## Goal

Add unit-level test coverage for `MoePermuteNopadFwdOp` (conformance + at-least-two-workloads), wire `source.kernel_map` in the manifest, fix any strict-parity drift, and flip the entry to `status: implemented`.

## Plan

1. Inspect the padded test file to mirror its structure for the nopad path (oracle = PyTorch reference of equivalent permute semantics).
2. Add nopad tests covering at least the two manifest workloads, in a sibling test module or as a parametrize block in `test_moe_permute.py` — whichever keeps the file under existing size budgets.
3. Add `source.kernel_map` in the manifest entry pointing at the existing kernel class.
4. Run `python scripts/validate_manifest.py --check-op MoePermuteNopadFwdOp --strict`; fix any remaining drift.
5. Flip `status: spec-only → implemented`.

## Constraints

- Tests must reference an external oracle (PyTorch reference or known-good baseline derivation), not values fabricated by the dev agent.
- Joint PR is acceptable here because the new code path (kernel route) needs new test coverage; see `docs/design/trust-model.md#issue-authoring-declaring-scope`.

## Acceptance Criteria

- [ ] Modified files pass unit tests
- [ ] At least 2 conformance tests exist for `MoePermuteNopadFwdOp`, each tied to a manifest workload
- [ ] `python scripts/validate_manifest.py --strict --check-op MoePermuteNopadFwdOp` exits 0
- [ ] `MoePermuteNopadFwdOp` carries `status: implemented` after the PR

### LINKED ISSUE #1475: [Test][MOE] add MoeGroupedGemmNopadFwdOp test and bench and flip to implemented
### Symptom / Motivation

`MoeGroupedGemmNopadFwdOp` is `status: spec-only`. The Op class and kernel exist (`tileops/ops/moe/grouped_gemm_nopad.py`, `tileops/kernels/moe/grouped_gemm_nopad.py`) but **neither tests nor benchmarks reference it** — the op is completely untested on `main`. Manifest also lacks `source.kernel_map` and `workloads`.

### Root Cause Analysis

The kernel was landed as MoE infrastructure for fused experts but never received its own parity test or roofline bench. The manifest entry was authored at spec-only and never advanced.

### Related Files

- `tileops/manifest/moe.yaml` (entry `MoeGroupedGemmNopadFwdOp`)
- `tileops/ops/moe/grouped_gemm_nopad.py`
- `tileops/kernels/moe/grouped_gemm_nopad.py`
- `tests/ops/` (new file: `test_moe_grouped_gemm_nopad.py`)
- `benchmarks/ops/` (new file: `bench_moe_grouped_gemm_nopad.py`)

## Goal

Stand up parity tests and a manifest-driven benchmark for `MoeGroupedGemmNopadFwdOp`, complete manifest fields, fix any strict-parity drift, and flip to `status: implemented`.

## Plan

1. Add 2+ representative `workloads` to the manifest entry (typical MoE token-routing shapes).
2. Add `source.kernel_map` pointing at the existing kernel class.
3. Write `tests/ops/test_moe_grouped_gemm_nopad.py` with at least one conformance test per manifest workload, using a PyTorch reference (batched matmul over expert groups) as oracle.
4. Write `benchmarks/ops/bench_moe_grouped_gemm_nopad.py` that loads workloads from the manifest (`load_workloads`) and reports `(latency, eval_roofline())`.
5. Run `python scripts/validate_manifest.py --strict --check-op MoeGroupedGemmNopadFwdOp`; fix remaining drift.
6. Flip `status: spec-only → implemented`.

## Constraints

- Joint PR (op-layer manifest + new test + new bench) is required because the new manifest fields and the new test cover code paths that today have zero coverage; see `docs/design/trust-model.md#issue-authoring-declaring-scope`.
- Workloads must be representative MoE token-routing shapes, not synthetic edge cases optimized for the kernel.
- Oracle in tests must be a PyTorch-expressible reference (e.g., per-expert `torch.matmul` over a permute index), not values produced by the kernel under test.

## Acceptance Criteria

- [ ] Modified files pass unit tests
- [ ] At least 2 conformance tests exist, each tied to a manifest workload
- [ ] Manifest entry has `workloads` (≥2), `source.kernel_map`, and `status: implemented`
- [ ] `python scripts/validate_manifest.py --strict --check-op MoeGroupedGemmNopadFwdOp` exits 0
- [ ] `benchmarks/ops/bench_moe_grouped_gemm_nopad.py` runs end-to-end on the manifest workloads

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1475

## Summary

- Add `tests/ops/test_moe_grouped_gemm_nopad.py` — 3 curated parity cases covering bf16/fp16, uniform + skewed expert distribution, and K-aligned + K-unaligned (TMA fast path + predicated path).
- Add `benchmarks/ops/bench_moe_grouped_gemm_nopad.py` — manifest-driven (`load_workloads`), runs the 4 DeepSeek-V3 workloads (gate-up / down × decode / prefill) and reports latency alongside `op.eval_roofline()`.
- Flip `MoeGroupedGemmNopadFwdOp` manifest entry to `status: implemented`, add `source.kernel_map`, and declare 4 representative workloads (per status-flip carve-out in `.claude/rules/manifest-trust-model.md`).
- Rename `self.N` / `self.K` → `self.n` / `self.k` in `MoeGroupedGemmNopadFwdOp` so the codegen-emitted `eval_roofline` body (per `docs/design/roofline.md` §4.4) can read the params by manifest name; no external code referenced the uppercase attributes.

## Test plan

- [x] AC1 — `tests/ops/test_moe_grouped_gemm_nopad.py`: 3 passed
- [x] AC2 — ≥2 conformance cases, each covering a manifest workload branch (gate-up + down GEMM + K-alignment); curated shapes per `.claude/domain-rules/testing-budget.md` (manifest workloads are consumed by the bench, not the test). Same precedent as #1487 (closes #1474).
- [x] AC3 — Manifest entry has 4 `workloads`, `source.kernel_map: MoeGroupedGemmNopadKernel`, and `status: implemented`.
- [x] AC4 — `python scripts/validate_manifest.py --strict --check-op MoeGroupedGemmNopadFwdOp` exits 0 (1 warning: shape-inference codegen pending for ops without `shape` on inputs; fires for 14+ other implemented ops, not a regression).
- [x] AC5 — `benchmarks/ops/bench_moe_grouped_gemm_nopad.py` runs end-to-end on the 4 manifest workloads.
- [x] Downstream smoke: `tests/ops/test_moe_experts_nopad.py -m smoke` — 15 passed (`FusedMoeExperts` + `MoEExpertsNopad` call this op).

## Benchmark

**Workloads** (manifest-driven, DeepSeek-V3 MoE shapes, E=256):

| Label                            | numel | n    | k    | dtype    |
| -------------------------------- | ----- | ---- | ---- | -------- |
| deepseek-v3-decode-gate-up       | 4096  | 4096 | 7168 | bfloat16 |
| deepseek-v3-prefill-gate-up      | 32768 | 4096 | 7168 | bfloat16 |
| deepseek-v3-decode-down          | 4096  | 7168 | 2048 | bfloat16 |
| deepseek-v3-prefill-down         | 32768 | 7168 | 2048 | bfloat16 |

End-to-end pass: 4/4. Detailed latency / roofline numbers will be exercised by the H100/H200 CI bench job.

**Command:** `PYTHONPATH="$PWD" python -m pytest benchmarks/ops/bench_moe_grouped_gemm_nopad.py -v`

### CODE REVIEW

File: benchmarks/ops/bench_moe_grouped_gemm_nopad.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        self.n = n
+        self.k = k
+        self.dtype = dtype
+
+    def gen_inputs(self):
+        torch.manual_seed(42)
+        dev = "cuda"
+        base = max(1, self.numel // self.num_experts)
+        sizes = torch.full((self.num_experts,), base, dtype=torch.int32, device=dev)
+        sizes[-1] = self.numel - base * (self.num_experts - 1)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply benchmarks/ops/bench_moe_grouped_gemm_nopad.py:43]: Valid edge case but out of scope. Manifest workloads have numel ≥ 4096 and num_experts = 256, so the numel < num_experts regime is unreachable here. Will fold the defensive branch into a follow-up if benchmarks introduce small-numel shapes.

File: tests/ops/test_moe_grouped_gemm_nopad.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        sizes[-1] = numel - base * (num_experts - 1)
+    elif distribution == "skewed":
+        sizes = torch.ones(num_experts, dtype=torch.int32, device=device)
+        extra = numel - num_experts
+        top_experts = max(1, num_experts // 5)
+        per_top = extra // top_experts
+        sizes[:top_experts] += per_top
+        sizes[0] += extra - per_top * top_experts
+    else:
+        raise ValueError(f"unknown distribution: {distribution}")
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_moe_grouped_gemm_nopad.py:43]: Valid edge case but out of scope. Manifest workloads have numel ≥ 4096 and num_experts = 256, so the numel < num_experts regime is unreachable here. Will fold the defensive branch into a follow-up if benchmarks introduce small-numel shapes.

File: tests/ops/test_moe_grouped_gemm_nopad.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    sizes_l = true_sizes.tolist()
+    offsets_l = true_offsets.tolist()
+    for e in range(num_experts):
+        size_e = sizes_l[e]
+        if size_e == 0:
+            continue
+        off_e = offsets_l[e]
+        a_e = a[off_e:off_e + size_e].to(torch.float32)
+        b_e = b[e].to(torch.float32)
+        c[off_e:off_e + size_e] = (a_e @ b_e.T).to(a.dtype)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_moe_grouped_gemm_nopad.py:82]: Valid edge case but out of scope. Manifest workloads have numel ≥ 4096 and num_experts = 256, so the numel < num_experts regime is unreachable here. Will fold the defensive branch into a follow-up if benchmarks introduce small-numel shapes.

File: benchmarks/ops/bench_moe_grouped_gemm_nopad.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply benchmarks/ops/bench_moe_grouped_gemm_nopad.py:43]: Valid edge case but out of scope. Manifest workloads have numel ≥ 4096 and num_experts = 256, so the numel < num_experts regime is unreachable here. Will fold the defensive branch into a follow-up if benchmarks introduce small-numel shapes.

File: tests/ops/test_moe_grouped_gemm_nopad.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        base = max(1, numel // num_experts)
+        sizes = torch.full((num_experts,), base, dtype=torch.int32, device=device)
+        sizes[-1] = numel - base * (num_experts - 1)
+    elif distribution == "skewed":
+        sizes = torch.ones(num_experts, dtype=torch.int32, device=device)
+        extra = numel - num_experts
+        top_experts = max(1, num_experts // 5)
+        per_top = extra // top_experts
+        sizes[:top_experts] += per_top
+        sizes[0] += extra - per_top * top_experts
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_moe_grouped_gemm_nopad.py:41]: Valid edge case but out of scope. Manifest workloads have numel ≥ 4096 and num_experts = 256, so the numel < num_experts regime is unreachable here. Will fold the defensive branch into a follow-up if benchmarks introduce small-numel shapes.

File: tileops/manifest/moe.yaml [hunk 5]
```diff
@@ -339,8 +344,10 @@ MoeGroupedGemmNopadFwdOp:
   source:
     kernel: tileops/kernels/moe/moe_grouped_gemm_nopad.py
     op: tileops/ops/moe/moe_grouped_gemm_nopad.py
-    test: tests/ops/test_moe_fused_moe.py
-    bench: benchmarks/ops/bench_moe_fused_moe.py
+    test: tests/ops/test_moe_grouped_gemm_nopad.py
```
Comment (Ibuki-wind) [inline tileops/manifest/moe.yaml]: MoeGroupedGemmNopadFwdOp `source.test` / `source.bench` are outside the status-flip carve-out; `.claude/rules/manifest-trust-model.md` allows this mixed promotion PR to edit only `status`, `source.kernel_map`, and promotion `workloads`, so these source-path edits widen the manifest contract without a manifest-only review → leave `source.test` / `source.bench` unchanged in this PR, or split those source-path updates into a separate manifest-only PR.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
Manifest trust-boundary issue needs a split before this can land.

File: tileops/manifest/moe.yaml (hunk 5 above)
Comment (lcy-seso) [reply tileops/manifest/moe.yaml]: Adopted. Reverted source.test/source.bench in aef489e; manifest paths restored to pre-PR values.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
The source fix is good, but the PR body still describes the old manifest-only state and stale benchmark wording; update it to match current HEAD before approval.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
The source fix is correct, but the PR body still includes process narration in the `Notes` block; remove that follow-up section so the body records only final-state facts.

**Review by Ibuki-wind — APPROVED** [review]
Clean — no issues.
